Examples/Io/Root/src/RootMaterialDecorator.cpp

```diff
@@ -68,129 +70,281 @@ ActsExamples::RootMaterialDecorator::RootMaterialDecorator(
+
+      // Only go on when you have all histograms
+      if (n and v and o and min and max and t and x0 and l0 and A and Z and
+          rho) {
+        // Get the number of bins
+        int nbins0 = t->GetNbinsX();
+        int nbins1 = t->GetNbinsY();
+
+        // The material matrix
+        Acts::MaterialSlabMatrix materialMatrix(
+            nbins1, Acts::MaterialSlabVector(nbins0, Acts::MaterialSlab()));
+
+        // We need binned material properties
+        if (nbins0 * nbins1 > 1) {
+          // Fill the matrix first
+          for (int ib0 = 1; ib0 <= nbins0; ++ib0) {
+            for (int ib1 = 1; ib1 <= nbins1; ++ib1) {
+              double dt = t->GetBinContent(ib0, ib1);
+              if (dt > 0.) {
+                double dx0 = x0->GetBinContent(ib0, ib1);
+                double dl0 = l0->GetBinContent(ib0, ib1);
+                double da = A->GetBinContent(ib0, ib1);
+                double dz = Z->GetBinContent(ib0, ib1);
+                double drho = rho->GetBinContent(ib0, ib1);
+                // Create material properties
+                const auto material =
+                    Acts::Material::fromMassDensity(dx0, dl0, da, dz, drho);
+                materialMatrix[ib1 - 1][ib0 - 1] =
+                    Acts::MaterialSlab(material, dt);
+              }
+            }
+          }
+
+          // Now reconstruct the bin untilities
+          Acts::BinUtility bUtility;
+          for (int ib = 1; ib < n->GetNbinsX() + 1; ++ib) {
+            size_t nbins = size_t(n->GetBinContent(ib));
+            Acts::BinningValue val = Acts::BinningValue(v->GetBinContent(ib));
+            Acts::BinningOption opt = Acts::BinningOption(o->GetBinContent(ib));
+            float rmin = min->GetBinContent(ib);
+            float rmax = max->GetBinContent(ib);
+            bUtility += Acts::BinUtility(nbins, rmin, rmax, opt, val);
+          }
+          ACTS_VERBOSE("Created " << bUtility);
+
+          // Construct the binned material with the right bin utility
+          sMaterial = std::make_shared<const Acts::BinnedSurfaceMaterial>(
+              bUtility, std::move(materialMatrix));
+
+        } else {
+          // Only homogeneous material present
+          double dt = t->GetBinContent(1, 1);
+          double dx0 = x0->GetBinContent(1, 1);
+          double dl0 = l0->GetBinContent(1, 1);
+          double da = A->GetBinContent(1, 1);
+          double dz = Z->GetBinContent(1, 1);
+          double drho = rho->GetBinContent(1, 1);
+          // Create and set the homogenous surface material
+          const auto material =
+              Acts::Material::fromMassDensity(dx0, dl0, da, dz, drho);
+          sMaterial = std::make_shared<const Acts::HomogeneousSurfaceMaterial>(
+              Acts::MaterialSlab(material, dt));
+        }
+      }
+      ACTS_VERBOSE("Successfully read Material for : " << geoID);
+
+      // Insert into the new collection
+      m_surfaceMaterialMap.insert({geoID, std::move(sMaterial)});
```

Comment:
This and the next block are very similar, would it make sense to move this into a function to avoid copying code?

Reply:
To be honest it look similar because both part are reading root file that contain material but there are so many small difference between the two that I am not 100% sure how a function could do both. But I will try to think about it.

Reply:
I think we can leave it as is ... that's root writing, so I think it doesn't have to be too clever.

Reply:
Yes, I tried to think about it last night but it probably would be too complicated for what this is worth. I think this can be merged as is.